src/monitor/group_state_machine.c

```diff
@@ -109,14 +109,42 @@ ProceedGroupState(AutoFailoverNode *activeNode)
 
 	if (primaryNode == NULL)
 	{
-		/* that's a bug, really, maybe we could use an Assert() instead */
-		ereport(ERROR,
-				(errmsg("ProceedGroupState couldn't find the primary node "
-						"in formation \"%s\", group %d",
-						formationId, groupId),
-				 errdetail("activeNode is %s:%d in state %s",
-						   activeNode->nodeName, activeNode->nodePort,
-						   ReplicationStateGetName(activeNode->goalState))));
+		AutoFailoverNode *otherNode = OtherNodeInGroup(activeNode);
```

Comment:
(Again, some questions to understand better, not necessarily asking for change at this point)

I think for the first time we end up with `no primary exists` case. Does that mean that we're losing our `service availability` principle for this short duration? 

A semi-related question: If the user follows `perform switchover && enable maintenance on the secondary` pattern to enable maintenance on the original primary, do you see any differences in terms of `service availability`?

Reply:
In terms of service availability this PR does not change much, we still need to stop the primary before promoting the standby as this is the only way to make sure we're not going to miss writes and have a split-brain situation.

In terms of state machine, the state maintenance could belong to either a primary or a secondary in the coding you reviewed and I agree it makes it harder to reason about. So I just introduced a new state `prepare_promotion` that only belongs to the primary, and I think that solves it.